### DISCUSSION TITLE
Confused about harbor.http.middleware.cors to solve Cross-Origin Request Blocked:  (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200

### DISCUSSION THREAD

**Discussion by jascco** [savonet/liquidsoap#3915] [Q&A] [answered]
I'm confused how to use harbor.http.middleware.cors()
Do you use it once when setting up server.harbor or is it needed for each specific harbor.http.register()

I am using nginx as my web server on Ubuntu 24.04, I have configured several nginx virtual sites including one only accessible from my local area network for security reasons. 

http://wildoat05.local:8011/mycastle.php

In mycastle.php I've created several AJAX javascripts which send commands to liquidsoap

http://wildoat05.local:8007/program

liquidsoap receives the command, executes it and sends a response. The response is blocked due to CORS header ‘Access-Control-Allow-Origin’ missing).

How do I use harbor.http.middleware.cors to fix it so I receive the response from liquidsoap? This is all new to me so any help is most appreciated.

Here is one of my AJAX javascript functions:
/* Following routine interacts with the web server and sends user's
* program choice to liquidsoap via AJAX, waits for a response then
* updates id=harborResponse and returns. */
function selectProgram(music) {
	alert('selectProgram choice=' + music);
	let xhr = new XMLHttpRequest;
	let url = new URL('http://wildoat05.local:8007/program');
	url.searchParams.set('choice', music);
	xhr.open('GET', url);
	xhr.responseType = 'text';
	xhr.send();
	xhr.onload = function() {
		if (xhr.status != 200) { // HTTP error
			document.getElementById("harborResponse").innerHTML = xhr.status;
			return;
		}
		document.getElementById("harborResponse").innerHTML = xhr.response;
	};
} // end of selectProgram()

Here is my liquidsoap script as a text file:
[castle-alpine.txt](https://github.com/savonet/liquidsoap/files/15284161/castle-alpine.txt)

**Comment by vitoyucepi** [comment, marked as answer]
Hi @jascco,

Your problem with CORS could be solved in several ways.
1. Create http endpoints in your PHP script. These endpoints will handle and filter client requests. Hide liquidsoap from users and only send http requests from PHP.
1. Serve liquidsoap on the same domain, for example in the subdirectory. Configure nginx to forward requests using `proxy_pass`.
1. Register the middleware `harbor.http.middleware.register(harbor.http.middleware.cors(origin="example.com"))`.

**Reply by jascco** [reply]
Hello @vitoyucepi,

Thank you for the prompt reply and for providing three suggestions - alternatives are always helpful.

liquidsoap ports 8006, 8007, 8008 are blocked by my router so they are only accessible from my local area network

I hadn't thought of using nginx proxy capability - good suggestion

As a test I registered the middleware - it took a couple of iterations to get the origin headers to match exactly - the Firefox browser console was very helpful

Thank you once again, I've marked your reply as answered.
